### ISSUE TITLE
Support following symbolic links to *files* as well

### ISSUE THREAD

**Issue by mirabilos** [codehaus-plexus/plexus-io#37]
From my analysis of the problem in https://github.com/codehaus-plexus/plexus-archiver/issues/160#issuecomment-756938000 changes in plexus-io are needed to support following symbolic links to *files* (not directories).

We can track the plexus-io part here.

**Comment by mirabilos** [comment]
I _think_ I’ve got it; for plexus-io even configurable (following the already extant follow-symlinks setting). I’ll provide a PR shortly.

**Comment by plamentotev** [comment]
While it may not be intuitive I think `PlexusIoFileResourceCollection` works as intended. This collection return a collection of files as they are on the file system. This includes symlinks. It is up to the caller to decide how to handle symlinks. The flag affect only dirs as it affects which files are going to be included. 

I think it would be better to keep the collection as it is, but while compressing Plexus Archiver should follow the sylink instead of adding it to the archive..

**Comment by mirabilos** [comment]
I don’t think so. It already includes a “follow symlinks” mode, which just doesn’t work “right”, it only follows symlinks to directories, which I haven’t seen _anywhere_ else.

I replied in more detail in https://github.com/codehaus-plexus/plexus-archiver/issues/160#issuecomment-757223767

**Comment by plamentotev** [comment]
Now that I had time to look again on the pull request it does make sense. I'm only concerned that it changes the default behavior of a code that existed for quite some time. Callers that haven't set the follow sym links flag at all will not get the sym links. That may break existing applications without them even knowing they are affected by the change.

**Comment by mirabilos** [comment]
Plamen Totev dixit:

>Now that I had time to look again on the pull request it does make
>sense. I'm only concerned that it changes the default behavior of a
>code that existed for quite some time. Callers that haven't set the

Yes, it’s probably a major version number bump, so that callers
can adapt. That doesn’t mean it should not be done, right? ☻

>follow sym links flag at all will not get the sym links. That may break
>existing applications without them even knowing they are affected by
>the change.

Right. Raising the major version number should make people look up
the reason in the release notes which would mention this.

I just hope that newer versions of the affected plugins will then
indeed make this configurable. Even discussing a similar issue in
https://issues.apache.org/jira/browse/MRESOURCES-237 made it clear
that the behaviour of following vs. copying the symlink as-is must
be configurable, as both have use cases that are equally valid.

bye,
//mirabilos
-- 
Infrastrukturexperte • tarent solutions GmbH
Am Dickobskreuz 10, D-53121 Bonn • http://www.tarent.de/
Telephon +49 228 54881-393 • Fax: +49 228 54881-235
HRB AG Bonn 5168 • USt-ID (VAT): DE122264941
Geschäftsführer: Dr. Stefan Barth, Kai Ebenrett, Boris Esser, Alexander Steeg
